reference/src/validity/unions.md

```diff
@@ -1,13 +1,232 @@
+
+<details><summary><b>Rationale</b></summary>
+
+The following examples must be supported, and therefore impose constraints on union value behaviour. The simplest solution, by far, is to treat the value representation of unions as merely being lists of bytes. While we do not discuss every possible angle here, it should be rapidly clear from just these two examples that any other example is significantly more complicated. Please trust us that every other alternatives that sounds even half-reasonable has been examined and has some critical flaw or another.
+
+</details>
+
+**We cannot require at least all fields, or even only one field, to be valid at all times.**
+
+```rust
+union Padding {
+    left: (Padded<u8>, u16), // [[u8, 1*pad], u16]
+    right: (u16, Padded<u8>), // [u16, [u8, 1*pad]]
+}
+let p = Padding{left: (0, 0)}; // resulting bytes: [0, uninit, 0, 0]
+p.right.1 = 1; // resulting bytes: [0, uninit, 1, uninit]
+fn f(_: Padding) {}
+f(p);
+```
+
+By the end of this example, the resulting union has no valid fields, because every field contains uninit, non-padding bytes. So therefore no field is even fully initialized, despite the fact that `p` was fully initialized as can be witnessed by the fact that the compiler allows it to be moved into `f`. This is all stable, Safe Rust, so these semantics being sound are a hard constraint.
+
+**Unions must preserve provenance.**
+
+```rust
+union Provenance<'a> {
+    raw: *const u32,
+    reference: &'a u32,
+}
+let x: u32;
+let u = Provenance(raw: &x);
+let y = unsafe{ *u.reference };
+```
+
+We must be able to carry provenance between the `raw` and `reference` fields in order for the assignment via `u.reference` to be valid. While this uses Unsafe Rust, this code is "obviously" sound. And therefore the union must be able to maintain provenance between the two pointers---even should the pointers be nested deeply within structs.
+
+[representation relation]: ../glossary.md#representation-relation
+[padding byte]: ../glossary.md#padding-byte
+
+### Niches
+
+## Valid values for #[repr(C)] and Raw-repr unions
+
+`#[repr(C)]` and [Raw-repr][raw repr] unions can take on any byte value.
+
+<details><summary><b>Rationale</b></summary>
+
+The purpose of the Raw-repr is to provide these semantics, which are easy to raeson about. Furthermore, C programmers are used to being able to treat unions like bags of bytes, more or less, and Rust programmers are similarly used to the same with `#[repr(C)]` unions. Therefore, they should both accept any arbitrary byte pattern.
+
+</details>
+
+## Possible niche values
+
+The presence of padding bytes, and writes to individual fields in general, makes niches hard to come by in unions. A niche representation of a union would have to not only be invalid for every single one of its fields, but also impossible to construct in Safe Rust with any combination of writes to any of its fields.
```

Comment:
Another reason why niches are hard to come by in a union: usually niches arise simply because the value model naturally constrains the set of valid representations. For unions we'd have to bolt something on to re-gain niches.

Reply:
Yeah, this is very true.

I'm increasingly inclined against it, even for `#[repr(transparent)]` unions with no ZSTs. That would make them really weird, and why not just use a struct then?

Reply:
(no longer part of this PR)